src/main/java/org/commcare/util/EncryptionKeyServiceProvider.java

```diff
@@ -0,0 +1,44 @@
+package org.commcare.util;
+
+import java.util.NoSuchElementException;
+import java.util.ServiceLoader;
+
+/**
+ * Utility class responsible for finding implementations of IEncryptionKeyProvider during runtime
+ * and loading them in memory
+ *
+ * @author avazirna
+ */
+
+public class EncryptionKeyServiceProvider {
+    private static EncryptionKeyServiceProvider serviceProvider;
+    private ServiceLoader<IEncryptionKeyProvider> loader;
+
+    private EncryptionKeyServiceProvider() {
+        loader = ServiceLoader.load(IEncryptionKeyProvider.class);
+    }
+
+    public static EncryptionKeyServiceProvider getInstance() {
+        if (serviceProvider == null) {
+            serviceProvider = new EncryptionKeyServiceProvider();
+        }
+        return serviceProvider;
+    }
+
+    public IEncryptionKeyProvider serviceImpl(boolean useKeyStoreIfAvailable) {
+        IEncryptionKeyProvider service = null;
+        if (loader.iterator().hasNext()) {
+            service = loader.iterator().next();
```

Comment:
This is a super interesting pattern, but I don't think it's the right one here.

What this code is really doing is determining if the system has a keystore provider available and using it if so, but there's no meaningful way to choose between multiple keystore-compatible providers, so it's just complex way of having two different things registered, and we should probably just have both in static memory from an initial registration if that's the case. 

This pattern is really for things like dependency injection from static libraries, and for doing fallback driven loading where, say, the code is looking for a keyprovider compatible with a specific algorithm/key type and it's going to choose the next one if it's not available.